src/dos/drive_fat.cpp

```diff
@@ -1021,24 +1043,27 @@ bool fatDrive::FileCreate(DOS_File **file, char *name, uint16_t attributes) {
 
 		/* Can we find the base directory? */
 		if(!getDirClustNum(name, &dirClust, true)) return false;
-		memset(&fileEntry, 0, sizeof(direntry));
+		fileEntry = {};
 		memcpy(&fileEntry.entryname, &pathName[0], 11);
-		fileEntry.attrib = (uint8_t)(attributes & 0xff);
+		fileEntry.attrib  = check_cast<uint8_t>(attributes);
+		fileEntry.modTime = DOS_GetSystemTimePacked();
+		fileEntry.modDate = DOS_GetSystemDatePacked();
 		addDirectoryEntry(dirClust, fileEntry);
 
 		/* Check if file exists now */
 		if(!getFileDirEntry(name, &fileEntry, &dirClust, &subEntry)) return false;
 	}
 
 	/* Empty file created, now lets open it */
-	/* TODO: check for read-only flag and requested write access */
 	*file = new fatFile(name, fileEntry.loFirstClust, fileEntry.entrysize, this);
-	(*file)->flags=OPEN_READWRITE;
-	((fatFile *)(*file))->dirCluster = dirClust;
-	((fatFile *)(*file))->dirIndex = subEntry;
-	/* Maybe modTime and date should be used ? (crt matches findnext) */
-	((fatFile *)(*file))->time = fileEntry.crtTime;
-	((fatFile *)(*file))->date = fileEntry.crtDate;
+	bool is_read_only    = fileEntry.attrib & DOS_ATTR_READ_ONLY;
+	(*file)->flags       = static_cast<uint32_t>(is_read_only ? OPEN_READ
+	                                                          : OPEN_READWRITE);
+	auto fat_file        = (fatFile*)(*file);
+	fat_file->dirCluster = dirClust;
+	fat_file->dirIndex   = subEntry;
+	fat_file->time       = fileEntry.modTime;
+	fat_file->date       = fileEntry.modDate;
```

Comment:
Ok; all that file-pointer and casting was quite a visual clutter.

I now see that we're actually creating `new fatFile` already, so it's very strange to see all this re-casting to get back to a `fatFile` type.

We can make it much clearer:

```c++
	const auto fat_file = new fatFile(name,
	                                  fileEntry.loFirstClust,
	                                  fileEntry.entrysize,
	                                  this);

	bool is_read_only    = fileEntry.attrib & DOS_ATTR_READ_ONLY;
	fat_file->flags      = is_read_only ? OPEN_READ : OPEN_READWRITE;
	fat_file->dirCluster = dirClust;
	fat_file->dirIndex   = subEntry;
	fat_file->time       = fileEntry.modTime;
	fat_file->date       = fileEntry.modDate;

	*file = fat_file; // assigned into member

	dos.errorcode = save_errorcode;
	return true;
```

Also note the casting on `flags` assignment is also gone here (`OPEN_READ : OPEN_READWRITE`).  If you can rebase against `main`, this has these flags now held in 8-bit unsigned enums.